networkx/algorithms/coloring/greedy_coloring.py

```diff
@@ -65,14 +65,47 @@ def strategy_smallest_last(G, colors):
 
     """
     H = G.copy(with_data=False)
-
-    def min_and_pop():
-        """Removes and returns the minimum degree node in ``H``."""
-        node = min(H, key=H.degree)
-        H.remove_node(node)
-        return node
-
-    return reversed([min_and_pop() for x in G])
+    result = []
+
+    # Build initial degree list (i.e. the bucket queue data structure)
+    degrees = defaultdict(set)  # set(), for fast random-access removals
+    for node, d in H.degree():
+        degrees[d].add(node)
+        try:
+            H.lbound = min(H.lbound, d)  # Lower bound on min-degree.
+        except AttributeError:
+            # Python 2.X doesn't support `nonlocal`.
+            # Workaround: set as attribute of the smallest-last graph.
+            H.lbound = d
+
+    def find_min_degree():
```

Comment:
In order to avoid global variables and avoid setting an attribute on the graph object, you could just inline this function (i.e. replace the body of this function where you compute `min_degree = find_min_degree()` below). If you do this, the code in the `for _ in G:` loop below will have four sections that implement the entire main algorithm: one for finding the minimum degree, one for popping a node of minimum degree, one for updating the neighbors, and one for removing the node. Then you won't even need the extra variable `lbound`, you will just use `min_degree` to store the minimum degree from the previous iteration of the loop (and set `min_degree` in the initialization code above instead of `lbound`).